sources/pam.c

```diff
@@ -0,0 +1,96 @@
+/*
+ * Odyssey.
+ *
+ * Scalable PostgreSQL connection pooler.
+*/
+
+#include <security/pam_appl.h>
+#include <kiwi.h>
+#include "pam.h"
+
+struct auth_request {
+    char *password;
+};
+
+
+static int
+pam_conversation(int msgc,
+                            const struct pam_message **msgv,
+                            struct pam_response **rspv,
+                            void *authdata) {
+
+    struct auth_request *request = (struct auth_request *) authdata;
+    int counter, retcode;
+
+    if (msgc < 1 || msgv == NULL || request == NULL) {
+        return PAM_CONV_ERR;
+    }
+    *rspv = malloc(msgc * sizeof(struct pam_response));
+    if (*rspv == NULL) {
+        return PAM_CONV_ERR;
+    }
+
+    memset(*rspv, 0, msgc * sizeof(struct pam_response));
+    retcode = PAM_SUCCESS;
+
+    for (counter = 0; counter < msgc; counter++) {
+        if (retcode != PAM_SUCCESS)
+            break;
+        switch (msgv[counter]->msg_style) {
+            case PAM_PROMPT_ECHO_OFF:
+                (*rspv)[counter].resp = strdup(request->password);
```

Comment:
When this gets freed?

Reply:
This gets freed somewhere deep inside pam library as i understand.